cocos/audio/android/AudioDecoder.cpp

```diff
@@ -0,0 +1,666 @@
+
+#include <android/log.h>
+#include <unistd.h>
+
+using namespace cocos2d;
+
+#define LOG_TAG "AudioDecoder"
+#define LOGD(...)  __android_log_print(ANDROID_LOG_DEBUG, LOG_TAG,__VA_ARGS__)
+#define LOGE(...)  __android_log_print(ANDROID_LOG_ERROR, LOG_TAG,__VA_ARGS__)
+
+/* size of the struct to retrieve the PCM format metadata values: the values we're interested in
+ * are SLuint32, but it is saved in the data field of a SLMetadataInfo, hence the larger size.
+ * Nate that this size is queried and displayed at l.452 for demonstration/test purposes.
+ *  */
+#define PCM_METADATA_VALUE_SIZE 32
+
+/* used to detect errors likely to have occured when the OpenSL ES framework fails to open
+ * a resource, for instance because a file URI is invalid, or an HTTP server doesn't respond.
+ */
+#define PREFETCHEVENT_ERROR_CANDIDATE (SL_PREFETCHEVENT_STATUSCHANGE | SL_PREFETCHEVENT_FILLLEVELCHANGE)
+
+//-----------------------------------------------------------------
+
+static void checkMetaData(int index, const char* key)
+{
+    if (index != -1) {
+        LOGD("Key %s is at index %d", key, index);
+    } else {
+        LOGE("Unable to find key %s", key);
+    }
+}
+
+class SLAudioDecoderCallbackProxy
+{
+public:
+    //-----------------------------------------------------------------
+    /* Callback for "prefetch" events, here used to detect audio resource opening errors */
+    static void prefetchEventCallback( SLPrefetchStatusItf caller,  void* context, SLuint32 event)
+    {
+        AudioDecoder* thiz = reinterpret_cast<AudioDecoder*>(context);
+        thiz->prefetchCallback(caller, event);
+    }
+
+    static void decPlayCallback(SLAndroidSimpleBufferQueueItf queueItf, void* context)
+    {
+        AudioDecoder* thiz = reinterpret_cast<AudioDecoder*>(context);
+        thiz->decodeToPcmCallback(queueItf);
+    }
+
+    static void decProgressCallback(SLPlayItf caller, void* context, SLuint32 event)
+    {
+        AudioDecoder* thiz = reinterpret_cast<AudioDecoder*>(context);
+        thiz->decodeProgressCallback(caller, event);
+    }
+};
+
+AudioDecoder::AudioDecoder(SLEngineItf engineItf, const std::string &url, int sampleRate)
+        : _engineItf(engineItf)
+        , _url(url)
+        , _playObj(nullptr)
+        , _pcmMetaData(nullptr)
+        , _formatQueried(false)
+        , _prefetchError(false)
+        , _counter(0)
+        , _numChannelsKeyIndex(-1)
+        , _sampleRateKeyIndex(-1)
+        , _bitsPerSampleKeyIndex(-1)
+        , _containerSizeKeyIndex(-1)
+        , _channelMaskKeyIndex(-1)
+        , _endiannessKeyIndex(-1)
+        , _eos(false)
+        , _sampleRate(sampleRate)
+        , _assetFd(0)
+{
+    memset(_pcmData, 0, sizeof(_pcmData));
+    auto pcmBuffer = std::make_shared<std::vector<char>>();
+    pcmBuffer->reserve(4096);
+    _result.pcmBuffer = pcmBuffer;
+}
+
+AudioDecoder::~AudioDecoder()
+{
+    if (_pcmMetaData != nullptr)
+    {
+        free(_pcmMetaData);
+        _pcmMetaData = nullptr;
```

Comment:
Should not skip this assignment since `AudioDecoder::start` may be interrupted and return false with `_pcmMetaData != null`, and the destructor should handle it.